File: 2217-step-by-step-directions-from-a-binary-tree-node-to-another/step-by-step-directions-from-a-binary-tree-node-to-another.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public String getDirections(TreeNode root, int startValue, int destValue) {
        // Find the LCA of the start and destination nodes
        TreeNode lca = findLCA(root, startValue, destValue);

        // Generate paths from the start node and destination node to the LCA
        StringBuilder pathToStart = new StringBuilder();
        StringBuilder pathToDest = new StringBuilder();
        findPath(lca, startValue, pathToStart);
        findPath(lca, destValue, pathToDest);

        // Convert the path to start node to 'U' steps
        for(int i=0; i<pathToStart.length(); i++){
            pathToStart.setCharAt(i, 'U');
        }

        return pathToStart.toString() + pathToDest.reverse().toString();
    }
    private TreeNode findLCA(TreeNode root, int p, int q){
        if(root == null){
            return null;
        }
        if(root.val == p || root.val == q){
            return root;
        }
        TreeNode left = findLCA(root.left, p, q);
        TreeNode right = findLCA(root.right, p, q);

        if(left != null && right != null){
            return root;
        }

        return left == null?right:left;
    }

    private boolean findPath(TreeNode root, int value, StringBuilder path){
        if (root == null) {
            return false;
        }
        if (root.val == value) {
            return true;
        }
        if (findPath(root.left, value, path)) {
            path.append('L');
            return true;
        }
        if (findPath(root.right, value, path)) {
            path.append('R');
            return true;
        }
        return false;
    }
}

File: 2217-step-by-step-directions-from-a-binary-tree-node-to-another/step-by-step-directions-from-a-binary-tree-node-to-another.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    string getDirections(TreeNode* root, int startValue, int destValue) {
        TreeNode* lca = findLCA(root, startValue, destValue);
        string pathToStart, pathToDest;
        findPath(lca, startValue, pathToStart);
        findPath(lca, destValue, pathToDest);

        // Convert pathToStart to 'U'
        for (char& c : pathToStart) {
            c = 'U';
        }

        reverse(pathToDest.begin(), pathToDest.end());

        return pathToStart + pathToDest;
    }
    
private:
    TreeNode* findLCA(TreeNode* root, int p, int q) {
        if (root == nullptr) {
            return nullptr;
        }
        if (root->val == p || root->val == q) {
            return root;
        }
        TreeNode* left = findLCA(root->left, p, q);
        TreeNode* right = findLCA(root->right, p, q);

        if (left != nullptr && right != nullptr) {
            return root;
        }

        return left != nullptr ? left : right;
    }

    bool findPath(TreeNode* root, int value, string& path) {
        if (root == nullptr) {
            return false;
        }
        if (root->val == value) {
            return true;
        }
        if (findPath(root->left, value, path)) {
            path.push_back('L');
            return true;
        }
        if (findPath(root->right, value, path)) {
            path.push_back('R');
            return true;
        }
        return false;
    }
};
